### PR TITLE
Pull Request description automation

### PR DESCRIPTION
**Pull request by michaelachrisco**
## Changes
1. Add Ruby based [Danger](https://danger.systems/ruby/) to the S&P repo 
2. Created Shift3PullRequestBot to create the messages on the fly as per Dangers instructions.
3. Added the `Dangerfile` a ruby based script that allows for certain messages to be auto-created when a Pull Request is made. 
4. `Bundle` now controls dependencies within the CI process instead of base `gem install`.  

## Purpose
When a pull request is created, we have a certain standard within the organization based on this documentation: https://github.com/Shift3/standards-and-practices/blob/main/standards/contributing.md 

We should have some sort of automated script that informs the developer what rules they may be unknowingly violating. It helps everyone by defining each one of these rules formally within the `Dangerfile`, sets us up for successful documentation/Pull Requests, and saves time by letting the developer know near instantly of what is expected in this repo. 

## Approach
This change adds a Danger file which contains all the PR specific formalized rules.:
1. All Pull Requests must have an issue associated with it.
2. PRs must have a Changes and Purpose statement contained within it.

I am intentionally making these _warnings_ at the moment since this is somewhat of an experiment. I want to see if these issues get resolved. I also left out the `Learning ` section since sometimes that is not relevant. These rules are up for discussion so please let me know if you have another opinion!

What the bot looks like when its working: 
image

To see an example of this working: 
https://github.com/Shift3/standards-and-practices/pull/239

## Learning
You can run danger locally without need of a CI with the below syntax.
`bundle exec danger pr https://github.com/Shift3/standards-and-practices/pull/238`
It runs in a matter of seconds so its not a huge strain on the CI process, but every little bit counts. 

Setup Docs: https://danger.systems/guides/getting_started.html

Scroll down to see the GitHub references: https://danger.systems/reference.html

Closes #238

### CODE REVIEW

File: Dangerfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Make it more obvious that a PR is a work in progress and shouldn't be merged yet
+warn("PR is classified as Work in Progress. When complete, please remove the [WIP] and assign reviewers.") if github.pr_title.include? "[WIP]"
```
Comment (Karvel) [inline Dangerfile:2]: Does this understand the GitHub Draft/Ready PR system?

Comment (Karvel) [reply Dangerfile:2]: Also, while this rule is useful, I don't know how much WIP is part of the GitHub PR culture.

File: Dangerfile [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+warn "Please link to a GitHub issue within this pull request. Example: Closes #9999" unless (github.pr_body.include? "Closes #" or github.pr_body.include? "References #")
+
+# Add in our standard using GitHub Closes within the Pull request body.
+warn "Please add a Changes section to your pull request description." unless github.pr_body.include? "Changes"
+warn "Please add a Purpose section to your pull request description." unless github.pr_body.include? "Purpose"
+
+# Warning if the user has not assigned themselves to the Pull Request.
+warn "This PR does not have any assignees yet. You may want to assign yourself if this is your Pull Request." unless github.pr_json["assignee"]
```
Comment (Karvel) [inline Dangerfile:11]: I would be consistent about "Pull Request" vs "pull request".

**Review by Karvel — COMMENTED** [review]
Some initial code review comments.

File: Dangerfile (hunk 1 above)
Comment (Karvel) [reply Dangerfile:2]: You might want to add "[wip]", "wip", and "WIP" if you can't do some equivalent of `toUpper()`.

**Review by soniaroh — APPROVED** [review]
this is great! thanks for making the PR opening process smooth and fault-free!

**Review by lizl714 — COMMENTED** [review]
Looks good - I agree with Elanna's comments

File: Dangerfile (hunk 1 above)
Comment (michaelachrisco) [reply Dangerfile:2]: As per our slack conversation, WIP branches are sparse and have no definitive standard within the realm of GitHub. Its up to the companies themselves to enforce. As such, I think leaving out this rule is better for now. However, the `toUpper()` comment makes sense and should be used within the rest of the Dangerfile.

File: Dangerfile (hunk 2 above)
Comment (michaelachrisco) [reply Dangerfile]: Thanks, good idea. 

Fixed via: 0eb3fe0b3868536112e78c5f68e0d6617a5e719f

File: Dangerfile (hunk 1 above)
Comment (michaelachrisco) [reply Dangerfile]: Removed via: 0eb3fe0b3868536112e78c5f68e0d6617a5e719f

**Comment by michaelachrisco** [comment]
#240 is preventing the CI job from moving forward. Created a hotfix for the issue in a separate PR.

File: Dangerfile [hunk 3]
```diff
@@ -1,11 +1,8 @@
-# Make it more obvious that a PR is a work in progress and shouldn't be merged yet
```
Comment (Karvel) [inline Dangerfile]: I get removing this because it's not consistent, but I do like the thinking behind it.

File: Dangerfile [hunk 4]
```diff
@@ -1,11 +1,8 @@
...
-warn("PR is classified as Work in Progress. When complete, please remove the [WIP] and assign reviewers.") if github.pr_title.include? "[WIP]"
-
-warn "Please link to a GitHub issue within this pull request. Example: Closes #9999" unless (github.pr_body.include? "Closes #" or github.pr_body.include? "References #")
+warn "Please link to a GitHub issue within this pull request. Example: Closes #9999" unless (github.pr_body.downcase.include? "closes #" or github.pr_body.downcase.include? "references #")
 
 # Add in our standard using GitHub Closes within the Pull request body.
-warn "Please add a Changes section to your pull request description." unless github.pr_body.include? "Changes"
-warn "Please add a Purpose section to your pull request description." unless github.pr_body.include? "Purpose"
+warn "Please add a Changes section to your Pull Request description." unless github.pr_body.downcase.include? "changes"
```
Comment (Karvel) [inline Dangerfile:4]: I'm glad there's an actual `.downcase`!

File: Dangerfile (hunk 3 above)
Comment (michaelachrisco) [reply Dangerfile]: Yep! I think we just need a standard and then we can revisit.

File: Dangerfile (hunk 4 above)
Comment (michaelachrisco) [reply Dangerfile:4]: Yep just some ruby. Small scripts like this are perfect for the lang.

File: Dangerfile (hunk 3 above)
Comment (kmausolf) [reply Dangerfile]: Perhaps we could encourage folks to use the Draft PR feature for these situations?

Comment (michaelachrisco) [reply Dangerfile]: That would be nice :). I am not sure what others options are on Draft PRs?

**Pull request merged**